### PR TITLE
Recover peers from translog, ignoring soft deletes

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today if soft deletes are enabled then we read the operations needed for peer
recovery from Lucene. However we do not currently make any attempt to retain
history in Lucene specifically for peer recoveries so we may discard it and
fall back to a more expensive file-based recovery. Yet we still retain
sufficient history in the translog to perform an operations-based peer
recovery.

In the long run we would like to fix this by retaining more history in Lucene,
possibly using shard history retention leases (#37165). For now, however, this
commit reverts to performing peer recoveries using the history retained in the
translog regardless of whether soft deletes are enabled or not.

### CODE REVIEW

**Review by dnhatn — APPROVED** [review]
I left some comments but LGTM. You need to remove or mute testRecoveryWithOutOfOrderDeleteWithSoftDeletes.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -2575,7 +2566,12 @@ public final long getMinRetainedSeqNo() {
     @Override
     public Closeable acquireRetentionLock() {
         if (softDeleteEnabled) {
-            return softDeletesPolicy.acquireRetentionLock();
+            final Closeable softDeletesRetentionLock = softDeletesPolicy.acquireRetentionLock();
+            final Closeable translogRetentionLock = translog.acquireRetentionLock();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Good point, and similarly on release. I pushed 7b6c2a0deb280e354d4a93b72ef2eff317995050.

File: server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java [hunk 2]
```diff
@@ -786,4 +788,55 @@ public void sendRequest(Transport.Connection connection, long requestId, String
             assertHitCount(client().prepareSearch(indexName).get(), numDocs);
         }
     }
+
+    @TestLogging("org.elasticsearch.indices.recovery:TRACE")
+    public void testHistoryRetention() throws Exception {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java]: No, this test motivates this change, in the sense that it fails on `master` but passes on this branch.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -2575,7 +2566,19 @@ public final long getMinRetainedSeqNo() {
...
         if (softDeleteEnabled) {
-            return softDeletesPolicy.acquireRetentionLock();
+            final Closeable softDeletesRetentionLock = softDeletesPolicy.acquireRetentionLock();
+            final Closeable translogRetentionLock;
+            try {
+                translogRetentionLock = translog.acquireRetentionLock();
+            } catch (Exception e) {
+                try {
+                    softDeletesRetentionLock.close();
+                } catch (IOException ioexception) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: nit: make softDeletesRetentionLock Releasable so we won't have to catch here.

**Pull request merged**